### PR TITLE
Add repository

### PR DESCRIPTION
**Pull request by EXAN7742**
+ add IRepository
+ add BaseRepository
* edit appsettings

### CODE REVIEW

File: ExadelBonusPlus.DataAccess/BaseRepository.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+using MongoDB.Bson;
+using MongoDB.Driver;
+
+namespace ExadelBonusPlus.Services.Models
+{
+    public abstract class BaseRepository<TModel> : IRepository<TModel> 
+        where TModel : Entity
+    {
+
+        private readonly IConfiguration _configuration;
```
Comment (myuk96) [inline ExadelBonusPlus.DataAccess/BaseRepository.cs:16]: Try to use Options pattern instead
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-5.0

File: ExadelBonusPlus.DataAccess/BaseRepository.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
```
Comment (EXAN7742) [inline ExadelBonusPlus.DataAccess/BaseRepository.cs:3]: test comment

**Review by myuk96 — CHANGES_REQUESTED** [review]

**Review by myuk96 — CHANGES_REQUESTED** [review]

**Review by myuk96 — CHANGES_REQUESTED** [review]

File: ExadelBonusPlus.DataAccess/BaseRepository.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            _mongoClient = new MongoClient(_mongoDbSettings.ConnectionString);
+            _database = _mongoClient.GetDatabase(_mongoDbSettings.DatabaseName);
+        }
+
+        private IMongoCollection<TModel> dbSet()
+        {
+            return _database.GetCollection<TModel>(nameof(TModel));
+        }
+
+        public virtual async Task<TModel> Add(TModel obj)
```
Comment (myuk96) [inline ExadelBonusPlus.DataAccess/BaseRepository.cs:30]: Remove all "async" statement  without "await" statement

**Review by myuk96 — CHANGES_REQUESTED** [review]

File: ExadelBonusPlus.DataAccess/BaseRepository.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            var result = dbSet().InsertOneAsync(obj);
+            return obj;
+        }
+        
+        public virtual async Task<IEnumerable<TModel>> GetAll()
```
Comment (myuk96) [inline ExadelBonusPlus.DataAccess/BaseRepository.cs:36]: Make all method that returns Task with Async postfix to show consumers it explicitly

**Review by myuk96 — CHANGES_REQUESTED** [review]

File: ExadelBonusPlus.DataAccess/BaseRepository.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        public virtual Task<TModel> AddAsync(TModel obj)
+        {
+            GetCollection().InsertOne(obj);
+            return  Task.FromResult(obj);
+        }
+        
+        public virtual Task<IEnumerable<TModel>> GetAllAsync()
+        {
+            return Task.FromResult(GetCollection().Find(Builders<TModel>.Filter.Empty).ToList() as IEnumerable<TModel>);
```
Comment (myuk96) [inline ExadelBonusPlus.DataAccess/BaseRepository.cs:31]: just return ToListAsync()

File: ExadelBonusPlus.DataAccess/BaseRepository.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+        public virtual Task<TModel> GetByIdAsync(Guid id)
+        {
+            return GetCollection().Find(Builders<TModel>.Filter.Eq("_id", id)).FirstAsync();
+        }
+
+        public virtual Task<TModel> UpdateAsync(Guid id, TModel obj)
+        {
+            GetCollection().ReplaceOne(Builders<TModel>.Filter.Eq("_id", id), obj);
+            return Task.FromResult(obj);
```
Comment (myuk96) [inline ExadelBonusPlus.DataAccess/BaseRepository.cs:42]: it's not proper
in this case we are not waiting ReplaceOne operation and make it Fire and Forget one , that's bad in this particular case, so let's do not returns models on repository levels than we update it or remove it

**Review by myuk96 — CHANGES_REQUESTED** [review]

File: ExadelBonusPlus.DataAccess/BaseRepository.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        public virtual Task<TModel> GetByIdAsync(Guid id, CancellationToken cancellationToken = default)
+        {
+            return GetCollection().Find(Builders<TModel>.Filter.Eq("_id", id)).FirstAsync(cancellationToken);
+        }
+
+        public virtual Task UpdateAsync(Guid id, TModel obj,  CancellationToken cancellationToken = default)
+        {
+            return GetCollection().ReplaceOneAsync(Builders<TModel>.Filter.Eq("_id", id), obj, new UpdateOptions { IsUpsert = true }, cancellationToken);
```
Comment (myuk96) [inline ExadelBonusPlus.DataAccess/BaseRepository.cs:41]: Is it necessary to add new document if document with the Id does not exist?

**Review by myuk96 — CHANGES_REQUESTED** [review]

**Pull request merged**